Try method match

NikitaAware opened the pull request:
Thank you for contributing to NullAway!

Please note that once you click "Create Pull Request" you will be asked to sign our [Uber Contributor License Agreement](https://cla-assistant.io/uber/NullAway) via [CLA assistant](https://cla-assistant.io/).

Before pressing the "Create Pull Request" button, please provide the following:

  - [ ] A description about what and why you are contributing, even if it's trivial.

  - [ ] The issue number(s) or PR number(s) in the description if you are contributing in response to those.

  - [ ] If applicable, unit tests.

msridhar commented on nullaway/src/main/java/com/uber/nullaway/NullAway.java:
```diff
@@ -607,6 +607,7 @@ private void checkForMethodNullMarkedness(MethodTree tree, VisitorState state) {
     }
   }
 
+  @SuppressWarnings({"UnusedVariable"})
```
remove

msridhar commented on nullaway/src/main/java/com/uber/nullaway/NullAway.java:
```diff
@@ -901,6 +902,7 @@ public Description matchMemberReference(MemberReferenceTree tree, VisitorState s
    * @param state visitor state.
    * @return discovered error, or {@link Description#NO_MATCH} if no error
    */
+  @SuppressWarnings({"UnusedVariable"})
```
remove

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -404,6 +404,87 @@ private Type.ClassType typeWithPreservedAnnotations(ParameterizedTypeTree tree)
+  // here we are checking if we should report an error on not considering the @Nullable annotations
+  // for overriden and
+  // overriding method return type should be consistent with the type parameter
+  public boolean DoNullabilityAnnotationsMatch(
```
This should be split into two separate things: (1) updating the overriddenMethodType for type variables, and (2) checking that type parameter nullability matches.  For (1) we should just update the `overriddenMethodReturnNullness` at the caller and re-use the logic there for checking if there is a mismatch with the overriding method.  (2) should be in a separate method, or maybe we can even just expose a lightweight wrapper around `compareNullabilityAnnotations`

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -404,6 +404,87 @@ private Type.ClassType typeWithPreservedAnnotations(ParameterizedTypeTree tree)
+   * node.getTarget().getReceiver().getBlock().getNodes().get(0).rhs.type.tsym member type (type of
+   * f1, sym for apply)
+   */
+  public static Nullness getActualAnnotation(
```
This is actually pretty much the exact logic we need for what I wrote above for (1).  Here let's take a `Tree` as the parameter rather than `MethodInvocationNode` so it's more re-usable

msridhar commented on nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessPropagation.java:
```diff
@@ -1011,6 +1012,25 @@ Nullness returnValueNullness(
+      if (node != null && config.isJSpecifyMode()) {
+        if (node.getTarget() != null && node.getTarget().getMethod() != null) {
+          if (node.getTarget().getMethod() instanceof Symbol.MethodSymbol) {
+            if (((Symbol.MethodSymbol) node.getTarget().getMethod()).owner.isInterface()) {
```
We should not need to check for an interface method.  The method could be in a class as well.

msridhar reviewed (commented):
some initial comments

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -495,4 +550,68 @@ public void compareNullabilityOfGenericParameters(
+    Type typeParamType =
+        state
+            .getTypes()
+            .memberType(overridingMethod.owner.type, overridingMethod.getReturnType().tsym);
```
Here shouldn't the second argument be the return type of the _overridden_ method, not overriding?

NikitaAware replied:
NO. I am checking if the annotations of Type parameter of that class match the annotations of the method  return type

msridhar replied:
I don't understand this check then.  We should discuss

msridhar replied:
We discussed in person.  We don't need this method.  We will add the return type check to `checkTypeParameterNullnessForOverridingMethodParameterType` and rename that method

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -495,4 +550,68 @@ public void compareNullabilityOfGenericParameters(
+        && overridingMethod.getReturnType() instanceof Type.ClassType)) {
+      return;
+    }
+    if (!(overridingMethod.getReturnType() instanceof Type.ClassType)) {
+      return;
+    }
```
this check is redundant

msridhar reviewed (commented):
I don't see any updates to handle generic parameter nullability; it would happen around here:

https://github.com/msridhar/NullAway/commit/0781cd8d28cd6bef430fa0d793078bd6403e8129#diff-a53186e00f6c045ebdbe4813d3e1ff0a469095e5034d5273571c09334a6cb4e4R721

Do you still intend to add that?  If you need a test case, let me know.  I wrote a couple minor comments on the current code, and I also pushed some cleanup and fixes.

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -498,4 +552,108 @@ public void compareGenericTypeParameterNullabilityForCall(
+      MethodTree tree, Symbol.MethodSymbol overridingMethod, Symbol.MethodSymbol overriddenMethod) {
+    List<? extends VariableTree> methodParameters = tree.getParameters();
+    List<Symbol.VarSymbol> overriddenMethodParams = overriddenMethod.getParameters();
+    Type ownerClassType =
+        state.getTypes().asSuper(overridingMethod.owner.type, overriddenMethod.owner);
+    if (ownerClassType == null) {
+      return;
+    }
```
I think this code is unnecessary?

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -498,4 +552,108 @@ public void compareGenericTypeParameterNullabilityForCall(
+    }
+    for (int i = 0; i < methodParameters.size(); i++) {
+      Type methodParameterType = getTreeType(methodParameters.get(i));
+      // p
```
?

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -498,4 +552,108 @@ public void compareGenericTypeParameterNullabilityForCall(
+    for (int i = 0; i < methodParameters.size(); i++) {
+      Type methodParameterType = getTreeType(methodParameters.get(i));
+      // p
+      Type typeParameterTypeOverriddenMethod =
+          state.getTypes().memberType(ownerClassType, overriddenMethodParams.get(i));
```
Rather than calling `memberType` for each overridden parameter type individually, why not call it once for the overridden method to get a `MethodType` and then get the parameter types from there?

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -498,4 +552,108 @@ public void compareGenericTypeParameterNullabilityForCall(
+        continue;
+      }
+      // type of type param corresponding to P
+      Type typeParameterType =
+          state.getTypes().memberType(ownerClassType, typeParameterTypeOverriddenMethod.tsym);
```
Why are you calling `memberType` again?

msridhar commented on nullaway/src/main/java/com/uber/nullaway/NullAway.java:
```diff
@@ -1662,6 +1673,11 @@ private Description handleInvocation(
       // This statement should be unreachable without assigning actual beforehand:
       Preconditions.checkNotNull(actual);
       // make sure we are passing a non-null value
+      if (config.isJSpecifyMode()) {
+        mayActualBeNull =
+            new GenericsChecks(state, config, this)
+                .hasMismatchedNullabilityOfArguments(methodSymbol, actualParams);
+      }
```
This is the wrong place to be modifying the logic.  You should be updating `argumentPositionNullness[i]` for each parameter around line 1633, and then the rest of the logic will just work.

NikitaAware replied:
will update the code. Thanks.

msridhar reviewed (commented):
Unfortunately there are still several problems

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -498,4 +553,88 @@ public void compareGenericTypeParameterNullabilityForCall(
+        tree, overridingMethod, overriddenMethod);
+  }
+
+  public Nullness getMethodParamNullness(VarSymbol param) {
```
The logic here isn't correct.  We're trying to handle a case like this from the test:
```java
Fn<@Nullable String, String> f1 = new TestFunc2();
// should get no error here
f1.apply(null);
```
I think the logic in this method sees that the parameter type is `P` and then looks at its upper bound and sees that it is `@Nullable Object`, and allows the call.  But that logic will also allow this call, which should _not_ be allowed:
```java
Fn<String, String> f1 = new TestFunc2();
// should get an error here!
f1.apply(null);
```
The right steps here are similar to the logic in `getOverriddenMethodReturnTypeNullness`.  For the above cases, we should get the type `R` of the receiver argument (`f1`), and then use `memberType()` to see what the method type of `apply()` would be if it were a member of `R`.  From that method type, you can correctly determine whether the relevant parameter should be treated as `@Nullable` or not.

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -498,4 +553,92 @@ public void compareGenericTypeParameterNullabilityForCall(
+      Type typeParameterType = typeParameterTypes.get(i);
+      if (typeParameterType instanceof Type.ClassType
+          && methodParameterType instanceof Type.ClassType) {
+        boolean hasNullableAnnotationTypeParamType =
+            Nullness.hasNullableAnnotation(
+                typeParameterType.getAnnotationMirrors().stream(), config);
+        boolean hasNullableAnnotationMethodParameterType =
+            Nullness.hasNullableAnnotation(
+                methodParameterType.getAnnotationMirrors().stream(), config);
+        if (hasNullableAnnotationTypeParamType != hasNullableAnnotationMethodParameterType) {
+          reportInvalidOverridingMethodParamTypeError(
+              methodParameters.get(i), typeParameterType, methodParameterType);
+        }
```
This is the wrong check, as it forces the nullability of the parameter in the overriding method to be the same as the overridden method.  But, if the overridden method has a `@NonNull` parameter, it is allowed for the overriding method to have a `@Nullable` parameter.  I have added a corresponding test case.

This code here should be deleted.  Instead, around line 728 in `NullAway.java`, we should add another call to `getMethodParamNullness()` so that `overriddenMethodArgNullnessMap[i]` is set correctly for code using generics in JSpecify mode.  We do want to keep the check below that the type param nullability annotations match.

NikitaAware replied:
Thanks for explaining. I was trying to compare the nullability annotations of the type parameter and the method parameter of the overriding method here but realized it is the wrong place to do so. I updated the code as per your suggestion.

XN137 committed:
refactor simple onOverrideMayBeNullExpr handlers (#747)

When `exprMayBeNull` is true, return immediately, to make clear that these handlers only support `false -> true` transitions

Co-authored-by: Manu Sridharan <msridhar@gmail.com>

nimakarimipour committed:
Update path serialization for class files (#752)

This PR partially resolves #716 by removing URI serialization for classes coming from `.class` files in all serializations.

In Annotator all serialized paths in downstream dependencies analysis phase, are neglected and retrieved from internal data structure of Annotator. Please see [Github](https://github.com/ucr-riple/NullAwayAnnotator/blob/4587c9821f4ccd152c1e6853a0cd10df5bcf277d/annotator-core/src/main/java/edu/ucr/cs/riple/core/cache/downstream/DownstreamImpactEvaluator.java#L85)

I tested Annotator with a local snapshot of NullAway with latest changes in this PR and I did not see any problem.

#716 is partially resolved since it requires updating path constructions in tests as well, which will be addressed in a separate PR.

XN137 committed:
NULL_LITERAL expressions may always be null (#749)

there is no reason to give handlers a chance to override this because we
are also not doing this for other literals that we filter out at the
start of the method

NikitaAware committed:
all changes for handling method overrides

XN137 committed:
refactor: streamline mayBeNullExpr flow (#753)

the uniform flow is now:
- return early for obvious expression types
- determine initial exprMayBeNull through simple checks
- apply handler overrides in single place
- finally run dataflow if exprMayBeNull is true in single place

XN137 committed:
refactor LibraryModelsHandler.onOverrideMayBeNullExpr (#754)

Also removes the call to `nullnessFromDataflow` from within this handler.

msridhar committed:
Merge branch 'master' into try-method-match

msridhar committed:
code for pretty printing generic types

msridhar committed:
better printing of types in errors related to generic types

msridhar committed:
fix compile error on JDK 8

msridhar committed:
Merge branch 'generic-type-pretty' into try-method-match

msridhar committed:
fix

msridhar committed:
cleanup

msridhar committed:
better error message

msridhar committed:
WIP

msridhar committed:
Merge branch 'master' into check-toolchains

msridhar committed:
test

msridhar committed:
reorder

msridhar committed:
just windows

msridhar committed:
Merge branch 'master' into generic-type-pretty

msridhar committed:
Merge branch 'check-toolchains' into generic-type-pretty

msridhar committed:
Merge branch 'generic-type-pretty' into try-method-match

msridhar committed:
Fix CI on Windows (#759)

msridhar committed:
Merge branch 'master' into generic-type-pretty

msridhar committed:
Merge branch 'generic-type-pretty' into try-method-match

msridhar committed:
tweaks

msridhar committed:
cleanup

msridhar committed:
more cleanup

msridhar committed:
cleanup

msridhar committed:
docs

msridhar committed:
more

msridhar committed:
more

msridhar committed:
more

msridhar committed:
more

msridhar committed:
cleanup tests

msridhar committed:
another rename

msridhar committed:
Clearer printing of types in errors related to generics (#758)

Previously, we would print fully-qualified names for all types, including annotation types. Now we use simple names for types, including types in annotations, which generally makes the message easier to read. We update some (but not all) test assertions to check how a type is printed.

msridhar committed:
Merge branch 'master' into try-method-match

XN137 committed:
Upgrade to Gradle 8.1.1 (#762)

upgraded by running twice:
./gradlew wrapper --gradle-version 8.1.1 --gradle-distribution-sha256-sum e111cb9948407e26351227dabce49822fb88c37ee72f1d1582a69c68af2e702f

see also:
https://gradle.org/release-checksums/
https://docs.gradle.org/8.1.1/release-notes.html

XN137 committed:
Upgrade to Error Prone 2.19.1 (#763)

need to fix a few new warnings:
- deprecated method usage
- not an actual javadoc comment

msridhar committed:
Merge branch 'master' into try-method-match

msridhar committed:
fix nullaway error

lazaroclapp committed:
Fix error inside Lombok generated code for @Nullable @Builder.Default (#765)

When given code such as:

```
@Builder
public class LombokDTO {
  @Nullable @Builder.Default private String fieldWithNullDefault = null;
}
```

Lombok internally generates the following method:

```
@java.lang.SuppressWarnings(value = "all")
@lombok.Generated()
private static String $default$fieldWithNullDefault() {
    return null;
}
```

which does not propagate `@Nullable` to the method's return type!

While this method is marked as `@Generated` code and `@SuppressWarnings("all")`, that does not suppress NullAway under all configurations. In fact, we sometimes want to check generated code (setters/getters for auto-annotation, for example), so just counting all Lombok Generated code as unannotated is not always the desired behavior (it's optional behavior, enabled by the `TreatGeneratedAsUnannotated=true` flag).

Instead, we want to internally and implicitly propagate the `@Nullable` annotation from `fieldWithNullDefault` to the generated `$default$fieldWithNullDefault()`.

We do this in two steps:

1. We modify our checking of return statements to allow handlers' existing overriding of method return nullability to be taken into account when deciding if `return [nullable expression];` should result in a NullAway error.
2. We add a new handler for fixing the Lombok nullability propagation, by detecting these `$default$foo()` methods and looking at the nullability of `foo` to determine if the method should also be `@Nullable`.

In addition to this, we can suggest a fix upstream in Lombok to propagate `@Nullable` to these `$default$` methods when present on the field, but this would not obviate the need for this PR, since we are keeping compatibility with multiple older Lombok releases internally.

Edit: Also, note that coverage on `LombokHandler` is bad in terms of unit tests, but that's because coveralls doesn't count `test-java-lib-lombok/` as a test case. We need to build with Lombok to exercise most of this handler, so we can't really exercise it in unit tests (vs integration) without significantly artificial workarounds (i.e. manually replicating the Lombok-generated code).

lazaroclapp committed:
Revert #762 (Grade upgrade) and add CI checking of publishing flows (#767)

A recent upgrade to our gradle version (#762) silently broke our
publishing flows.

In particular, `publish` and `publishToMavenLocal` are broken on Java 8,
resulting in the following error:

```
Execution failed for task ':annotations:publishMavenPublicationToMavenLocal'.
> Failed to publish publication 'maven' to repository 'mavenLocal'
   > Invalid publication 'maven': artifact file does not exist: '/home/runner/work/NullAway/NullAway/annotations/build/publications/maven/module.json.asc'
```

See this [CI job
run](https://github.com/uber/NullAway/actions/runs/5261437721/jobs/9509542067?pr=767).
There are further errors when the release is signed.

Below, we revert the gradle upgrade, and also add a CI job running
`publishToMavenLocal` to prevent
future regressions (We can't trivially test signing these publications
in a secure manner, unfortunately,
so we skip that for now. If that specifically breaks in the future, we
could provide a test gpg key for signing
in CI).

msridhar committed:
Merge branch 'master' into try-method-match

msridhar committed:
Merge branch 'master' into try-method-match

msridhar committed:
fix class ordering in test

msridhar committed:
bug fix: did not handle direct dereference of method call expression with generic @Nullable return type

msridhar committed:
test interactions with @Contract

msridhar committed:
simplify using new API

msridhar committed:
better variable names

msridhar committed:
rename variables and add a TODO

msridhar committed:
add a comment

Asapin committed:
Added support for Apache Validate (#769)

Added support for Apache Validate:
https://github.com/apache/commons-lang/blob/master/src/main/java/org/apache/commons/lang3/Validate.java

We recently started to use these Apache methods to validate some data
but quickly found out that NullAway doesn't support them since they are
not properly annotated. Here's a fix that adds support for these methods
along with test cases for each method.

I noticed that Apache library generally has different validation methods
for String, Map and Collection, but they all end up with the same
signature when analyzed by NullAway (probably because these methods use
generics).
Examples of such methods:
```java
public static <T extends Collection<?>> T notEmpty(final T collection) {
}

public static <T extends Map<?, ?>> T notEmpty(final T map) {
}

public static <T extends CharSequence> T notEmpty(final T chars) {
}
```

But I decided to add a separate unit test for each of these methods in
case NullAway will change how it generates method's signature in the
future.

lazaroclapp committed:
Introduce FluentFutureHandler as a workaround for Guava Futures/FluentFuture (#771)

We need this due to our incomplete support for generics, and the fact
that #765 fixes a previous
false negative which would cause our `@NonNull` model for Guava's
`Function` to be ignored
when determining the correct overriding of the functional interface by a
lambda.

After the fix, however, code such as:

```
FluentFuture
            .from(...)
            .transform(s -> { if(...) {...} else { return null; } }, executor);
```

will fail with an error about `Function::apply` having a `@NonNull`
result. Where nullability should
actually depend on the parameter `T` of the `ListenableFuture<T>` in
question.

Unfortunately, usage of this futures API is common internally for us,
and our generics support is far
from fully supporting this in a sound manner, so this diff introduces a
(hopefully temporary) workaround,
in the form of handler for the Futures/FluentFuture Guava APIs:

This works by special casing the return nullability of
`com.google.common.base.Function` and
`com.google.common.util.concurrent.AsyncFunction` to be e.g.
`Function<@Nullable T>` whenever
these functional interfaces are implemented as a lambda expression
passed to a list of specific
methods of `com.google.common.util.concurrent.FluentFuture` or 
`com.google.common.util.concurrent.Futures`. This is unsound, but
permits the code example above to
pass NullAway checking, while strictly reducing the safety hole of
NullAway versions prior to PR #765.

msridhar committed:
Merge branch 'master' into try-method-match

lazaroclapp committed:
Prepare for release 0.10.11.

lazaroclapp committed:
Upgrade maven publish plugin to 0.21.0 and restore publishing (#773)

Without this change, the recent Gradle upgrade (#743) resulted in the
following error while trying to publish to maven central:

```
> Task :jar-infer:jar-infer-cli:publishShadowPublicationToLocalRepository FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Some problems were found with the configuration of task ':jar-infer:jar-infer-cli:publishShadowPublicationToLocalRepository' (type 'PublishToMavenRepository').
  - Gradle detected a problem with the following location: '/Users/lazaro/Uber/NullAway/jar-infer/jar-infer-cli/build/libs/jar-infer-cli-0.10.11.jar.asc'.
    
    Reason: Task ':jar-infer:jar-infer-cli:publishShadowPublicationToLocalRepository' uses this output of task ':jar-infer:jar-infer-cli:signMavenPublication' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':jar-infer:jar-infer-cli:signMavenPublication' as an input of ':jar-infer:jar-infer-cli:publishShadowPublicationToLocalRepository'.
      2. Declare an explicit dependency on ':jar-infer:jar-infer-cli:signMavenPublication' from ':jar-infer:jar-infer-cli:publishShadowPublicationToLocalRepository' using Task#dependsOn.
      3. Declare an explicit dependency on ':jar-infer:jar-infer-cli:signMavenPublication' from ':jar-infer:jar-infer-cli:publishShadowPublicationToLocalRepository' using Task#mustRunAfter.
```

This change fixes that issue and also upgrades
`com.vanniktech:gradle-maven-publish-plugin` to 0.21.0, the latest
compatible version (versions from 0.23.0 onwards don't seem to be
compatible with JDK8, and 0.22.0 gives a separate error).

Long term, we need to update the CI testing flow for publishing we added
in #768 to also cover jar signing workflows.

lazaroclapp committed:
Prepare next development version.

lazaroclapp committed:
Fix Publish Snapshot CI job (#774)

The name of these two configuration options was changed in newer
versions of `gradle-maven-publish-plugin`.

Without this change, #773 breaks our snapshot publishing GitHub Actions
CI workflow.

msridhar committed:
Add step to create release on GitHub (#775)

Users look for GitHub Releases these days (and can be easily notified of
such releases), so we should support them. I created a release for
0.10.11:

https://github.com/uber/NullAway/releases/tag/v0.10.11

The GitHub Release page has a nice feature for generating release notes
from commits, so eventually we may want to deprecate `CHANGELOG.md`.

msridhar committed:
Build the Android sample app on JDK 17 (#776)

Seems that this works now, so might as well run it on CI. The JarInfer
tasks still don't work on JDK 17 due to a WALA issue; I'll see if I can
work on that.

msridhar committed:
Update to Error Prone 2.20.0 (#772)

Pretty straightforward update

msridhar committed:
Merge branch 'master' into try-method-match

msridhar committed:
Add tasks to run JDK 8 tests on JDK 11+ (#778)

This PR adds tasks to additionally run certain test suites on JDK 8,
even when we are building on a JDK 11+ JVM. The approach is described
[here](https://jakewharton.com/build-on-latest-java-test-through-lowest-java/).
This moves us towards only supporting building NullAway on JDK 11+,
while still targeting Java 8 (so NullAway can still be used on JDK 8).
This will help us with keeping build tooling and dependencies up to
date, as more and more tools and plugins do not support running on JDK
8.

Specifically, we run the core NullAway tests and the tests from
`guava-recent-unit-tests` on JDK 8. We do _not_ run JarInfer tests, as
we intend to remove the ability to run JarInfer on JDK 8 (see #777).
Also, we only create the relevant test tasks when the Error Prone
version being used is 2.4.0 (compatible with Java 8). We have one CI
job, Java 11 + EP 2.4.0, that exercises this scenario. I confirmed
locally that when running the `testJdk8` tasks, the tests are running on
a JDK 8 JVM.

Additionally, the PR contains a minor tweak that allows
`publishToMavenLocal` to work when Gradle is run on JDK 11+.

msridhar committed:
Update to WALA 1.6.1 and remove ability to build on JDK 8 (#777)

WALA now runs on JDK 17. So, with this update, we can run our full build
on JDK 17, including JarInfer. Since WALA has required JDK 11 since
version 1.6.0, JarInfer now requires a JDK 11 VM to run, and will no
longer run on JDK 8. In order to get everything to build with this new
dependence, we have removed support for building on JDK 8. NullAway
itself will still run as part of a JDK 8 build (and it still tested
there, see #778), but running JarInfer now requires JDK 11. Much of this
PR is cleanup enabled by removing support for building on JDK 8.

msridhar committed:
Merge branch 'master' into try-method-match

msridhar committed:
Switch to Spotless for formatting Java code (#780)

msridhar committed:
Fix compatibility issue when building on JDK 17 but running on JDK 8 (#779)

Error Prone ran into a similar issue in
https://github.com/google/error-prone/issues/3895. We re-use the same
fix, and also add a CI job to test this case. We could avoid this
problem by only building releases on JDK 11, but might as well fix to
avoid that requirement.

msridhar committed:
Merge branch 'master' into try-method-match

armughan11 committed:
Added GCP JMH Benchmark Workflow  (#770)

Adds a workflow that is triggered via a "/benchmark" comment on a pull
request. Clones the main and PR branch on the GCP instance and runs the
`./gradlew jmh` on each one of them, returning the results as a comment
on the pull request. 
---------

Co-authored-by: Manu Sridharan <msridhar@gmail.com>

msridhar committed:
Set concurrency for JMH benchmarking workflow (#784)

We should only allow one benchmarking workflow to be running at any
given time, since we only have one VM allocated for it.

msridhar committed:
Merge branch 'master' into try-method-match

msridhar committed:
Disable daemon when running benchmarks (#786)

This might reduce the noise a bit when running benchmarks (since old
daemon processes won't be running).

msridhar committed:
Update to Gradle 8.2.1 (#781)

We had to back out a previous Gradle upgrade, but I believe now things
have been fixed up enough that we can upgrade again. One issue with the
upgrade is that publishing a build will fail if signing is not
functional and the version is not a `SNAPSHOT`. In particular, with our
default config, `LOCAL` versions must be signed. (This was the
documented behavior before this Gradle upgrade but for some reason the
build wasn't failing.) So, publishing `LOCAL` builds without signing
requires disabling a setting via a Gradle property or env var. We update
the CI config and the docs to reflect this.

I tested publishing a release to Sonatype (without pushing to Maven
Central) from my laptop and it succeeded. I have not yet validated the
actual artifacts. Will hold off on landing this until we're sure it's
thoroughly tested.

Fixes #768

msridhar committed:
JSpecify: avoid crashes when encountering raw types (#792)

We check for the presence of a raw type and (for now) bail out of any
checking. Adding further support for JSpecify checking of raw types is a
low priority (not even sure anything needs to be done).

This fixes the initial crash from #791, though we may want to leave that
issue open for now until we split off issues for the other crashes as
needed.

msridhar committed:
Fix JDK compatibility issue in LombokHandler and introduce AstHelpersBackports (#795)

Fixes #794. Since the JDK interface change was made for JDK 17, I think
even if NullAway was built on and targeted JDK 11, we would still have
this problem. I added some hacky unit test coverage for this code so
we'd have a way to test for regressions in the future. (The new unit
test fails the `:nullaway:testJdk8` task without this change.)

We missed this problem earlier since we suppressed a warning from
`AstHelpersSuggestions`. Even though we cannot accept some of the
suggestions from that check (since we may not require a recent enough
version of Error Prone), the issues it flags can be serious. So, this PR
removes all other suppressions of that checker and fixes the warnings.
We introduce an `AstHelpersBackports` class to contain any backported
logic from `AstHelpers` to enable the fixes.

msridhar committed:
Fix off-by-one error in JSpecify checking of parameter passing (#793)

Not sure how we missed this before...

lazaroclapp committed:
Improve auto-fixing of unnecessary castToNonNull calls (#796)

In particular, this allows removing unnecessary casts where the
castToNonNull method is specified by a library model, rather than a CLI
argument.

lazaroclapp committed:
Prepare for release 0.10.12.

lazaroclapp committed:
Prepare next development version.

msridhar committed:
Merge branch 'master' into try-method-match

msridhar committed:
Bump versions for some dependencies (#800)

To keep up to date and avoid surprises later.

msridhar committed:
Update to WALA 1.6.2 (#798)

We use a [new WALA feature](https://github.com/wala/WALA/pull/1286) to
only include the `java.base` module in the `AnalysisScope` used for
JarInfer. JarInfer does not require all JDK standard library modules to
be part of the analysis scope, just enough to make a valid class
hierarchy, and `java.base` suffices. This should speed up runs of
JarInfer, and in particular it speeds up our JarInfer regression tests.

msridhar committed:
Update to Error Prone 2.21.1 (#797)

Mostly straightforward. Error Prone has a new
[NullableOptional](https://errorprone.info/bugpattern/NullableOptional)
check that warned on a couple instances in our code base. I changed
these cases to use an `@Initializer` method which I think better
captures what is going on. Will update the required CI checks before
landing.

msridhar committed:
Merge branch 'master' into try-method-match

msridhar committed:
clean up check

msridhar committed:
rename type parameter

msridhar committed:
add true negative test

msridhar committed:
add true negative test

msridhar committed:
fix nullaway error

msridhar committed:
Add test for false negative

msridhar committed:
rewrite condition for clarity

msridhar committed:
refactor for clarity

msridhar commented:
Closing as this has all been merged upstream
